### PR TITLE
E2.02/fill html template

### PR DESCRIPTION
**Pull request by Steffengreiner**
**Description of changes**
Provides the method to fill the HTML Template with the List<NotificationContent> DTOs provided by the Use case.

### CODE REVIEW

File: sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/DependencyManager.groovy
Comment (jenniferboedker) [inline sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/DependencyManager.groovy:71]: why do you provide the notification list via the constructor and not trigger it via method call during runtime (eg in the presenter)

File: sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/CreateNotificationPresenter.groovy
Comment (jenniferboedker) [inline sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/CreateNotificationPresenter.groovy:16]: why is this a class variable? I think this should be like this because you dont have any advantage (or do I miss it?) to store it. It should only be used within the context of the createNotification() method :)

File: sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/EmailGenerator.groovy
Comment (jenniferboedker) [inline sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/EmailGenerator.groovy:17]: NIIIIIIt 🤓 
```suggestion
 * @since: 1.0.0
```

File: sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/EmailGenerator.groovy
Comment (jenniferboedker) [inline sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/EmailGenerator.groovy:40]: ok, keep in mind this should be the email that is provided and not your hardcoded mail. but I understand you did it for testing reasons :)

File: sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/EmailGenerator.groovy
Comment (jenniferboedker) [inline sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/EmailGenerator.groovy:28]: I would expect it to have a fillTemplate(List notifications) method rather than do it like this

File: sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/EmailGenerator.groovy
Comment (jenniferboedker) [inline sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/EmailGenerator.groovy:46]: why is that not a classvariable? this will be the same for this document, right?

File: sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/EmailGenerator.groovy
Comment (jenniferboedker) [inline sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/EmailGenerator.groovy:25]: maybe rename to make more apparent that these are the final email (are they?)

```suggestion
    private List<Document> emails = []
```

File: sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/EmailGenerator.groovy
Comment (jenniferboedker) [inline sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/EmailGenerator.groovy:50]: This would make more apparent what happens. I had difficulties understanding what happens here

```suggestion
        Document filledEmail = emailHTMLTemplate.fillTemplate(notificationContent)
        emails.add(filledEmail)
```

File: sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/EmailGenerator.groovy
Comment (jenniferboedker) [inline sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/EmailGenerator.groovy:54]: yaiii html 🙌🏼

File: sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/EmailGenerator.groovy
Comment (jenniferboedker) [inline sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/EmailGenerator.groovy:53]: is it still the strin gor the notification content DTO that you get here?

File: sample-notificator-domain/src/main/groovy/life/qbic/business/notification/create/CreateNotification.groovy
Comment (jenniferboedker) [inline sample-notificator-domain/src/main/groovy/life/qbic/business/notification/create/CreateNotification.groovy:34]: :D I like your examples

Comment (jenniferboedker) [reply sample-notificator-domain/src/main/groovy/life/qbic/business/notification/create/CreateNotification.groovy:34]: please delete it before merging to the use case :)

**Review by jenniferboedker — CHANGES_REQUESTED** [review]
Looks good so far, but sadly I cant test it. But I believe you if you say it runs :)

I gave you some feedback where the injection of values was not clear to me maybe we can quickly sync? Thanks again for putting so much work into this, I can see it!

File: sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/EmailGenerator.groovy
Comment (Steffengreiner) [reply sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/EmailGenerator.groovy:40]: I'll adapt this before merging thanks for reminding me 👍

File: sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/EmailGenerator.groovy
Comment (Steffengreiner) [reply sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/EmailGenerator.groovy:28]: Oh sure that's a good Idea!

File: sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/EmailGenerator.groovy
Comment (Steffengreiner) [reply sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/EmailGenerator.groovy:55]: 🤓 🌐

File: sample-notificator-domain/src/main/groovy/life/qbic/business/notification/create/CreateNotification.groovy
Comment (Steffengreiner) [reply sample-notificator-domain/src/main/groovy/life/qbic/business/notification/create/CreateNotification.groovy:34]: Will do thanks for the headsup!

File: sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/EmailGenerator.groovy
Comment (Steffengreiner) [reply sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/EmailGenerator.groovy:46]: True! However, The InputStream is consumed every time it's parsed, therefore it has to be generated anew for each notification.

File: sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/EmailGenerator.groovy
Comment (Steffengreiner) [reply sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/EmailGenerator.groovy:54]: Good point, i clarified it by putting the emailbody into a seperate map combined with the customer email address to avoid confusion

File: sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/EmailGenerator.groovy
Comment (jenniferboedker) [reply sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/EmailGenerator.groovy]: Is it? I don't think so. You write the input into the emailHTMLTemplate variable. I just checked the OM. By making it a class variable you do not only save time but also unnecessary file reading. Furthermore, your method would look cleaner (the method where you call this). 

I even think its better to store the URI rather than a stream. In the OM we had
```
  private static final URI OFFER_HTML_TEMPLATE_URI = OfferToPDFConverter.class.getClassLoader()
            .getResource("offer-template/offer.html").toURI()
```
as a class variable and read the info into the doc like this (also a class variable) but in your case it would make more sense to do your call as you did already.

`    private static final OfferTemplate TEMPLATE = new OfferTemplate(Jsoup.parse(new File(OFFER_HTML_TEMPLATE_URI),"UTF-8"))`

File: sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/EmailGenerator.groovy
Comment (jenniferboedker) [inline sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/EmailGenerator.groovy:31]:
```suggestion

```

File: sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/EmailGenerator.groovy
Comment (jenniferboedker) [inline sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/EmailGenerator.groovy:42]:
```suggestion
    void initializeEmailSubmission(List<NotificationConten> notificationContentList) {
        notificationContentList.each { NotificationContent notificationContent ->
            accessEmailTemplate()
            prepareHTMLEmail(notificationContent)
        }
        sendEmail()
```

File: sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/EmailGenerator.groovy
Comment (jenniferboedker) [inline sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/EmailGenerator.groovy:55]:
```suggestion
    private void sendEmails() {
```

File: sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/EmailGenerator.groovy
Comment (jenniferboedker) [inline sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/EmailGenerator.groovy:25]: why did that change?

Comment (jenniferboedker) [reply sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/EmailGenerator.groovy:25]: I see, because you changed the way you iterate over the content

File: sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/EmailGenerator.groovy
Comment (Steffengreiner) [reply sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/EmailGenerator.groovy]: True but apparently loading a resource from a jar is only possible via an Inputstream which is then consumed. I wasn't able to get it to work with an URI :(

File: sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/EmailGenerator.groovy
Comment (jenniferboedker) [inline sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/EmailGenerator.groovy:47]: Major nit
```suggestion
        Document filledEmailTemplate = emailHTMLTemplate.fillTemplate(notificationContent)
```

File: sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/EmailGenerator.groovy
Comment (jenniferboedker) [inline sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/EmailGenerator.groovy:48]:
```suggestion
        emails.put(filledEmailTemplate, notificationContent.customerEmailAddress)
```

**Review by jenniferboedker — APPROVED** [review]
Thanks for being so patient and putting some much work into this 🚀

**Pull request merged**
